### ISSUE TITLE
Corpus: Download and integrate Lingvaj Respondoj by Zamenhof

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#519]
## Parent Issue

Part of #518 - Tier 0 corpus expansion

## What is Lingvaj Respondoj?

**Linguistic Answers/Responses by L.L. Zamenhof (1889-1912)**

Zamenhof's authoritative answers to grammar and usage questions, published in early Esperanto journals (La Esperantisto, La Revuo, La Oficiala Gazeto).

**Content**:
- Direct linguistic guidance from Esperanto's creator
- Edge cases and grammatical ambiguities
- Authoritative rulings on disputed usages
- Historical context for grammar rules

**Why Tier 0**: Definitive authority - Zamenhof's own words on how Esperanto should work

## Download Sources

**Primary** (Recommended):
- [Internet Archive - Complete Collection](https://archive.org/details/kpbc.umk.pl.Gromadzenie_POPC_024_48_247196)
  - Complete collection (plena kolekto)
  - Multiple formats available
  - Public domain

**Alternative**:
- [PDF version (americanroads.us)](https://www.americanroads.us/esperanto/Lingvaj_Respondoj_Zamenhof.pdf)
- [Tekstaro.com version](https://tekstaro.com/t?nomo=lingvaj-respondoj) - 7th edition by Waringhien

## License

✅ **Public Domain** - Zamenhof died 1917 (>100 years ago)

## Estimated Size

- ~200 pages
- ~1,000-1,500 sentences (estimated)
- ~50-100 KB text

### 1. Parser Validation (#513-#516) - **CRITICAL**

Lingvaj Respondoj explains **how to parse edge cases**:

Example: If Zamenhof says "word X should be parsed as Y", that's the ground truth.

**Test suite**:
```python
def test_lingvaj_respondoj_examples():
    """Zamenhof's examples are definitive - parser MUST handle them."""
    
    # Load examples from Lingvaj Respondoj
    examples = load_lingvaj_respondoj_examples()
    
    failures = []
    for example in examples:
        ast = parse_sentence(example['text'])
        if ast['parse_statistics']['success_rate'] < 1.0:
            failures.append({
                'text': example['text'],
                'zamenhof_guidance': example['explanation'],
                'parser_result': ast
            })
    
    # Tier 0 - must be 100% correct
    assert len(failures) == 0, f"Parser failed on {len(failures)} Zamenhof examples"
```

### 2. Ambiguity Resolution (#516)

When multiple parses are possible, Zamenhof's answers show **which is correct**:
- Example: "eksano" - Zamenhof may explain whether it's "ek-sano" or "eks-ano"
- Guides disambiguation ranking

### 3. High-Quality Training Data

- **Root embeddings**: Co-occurrence patterns from authoritative source
- **M1 semantic**: Correct semantic usage examples
- **AST reasoning core**: Foundation patterns

### Step 1: Create Acquisition Script

**File**: `scripts/acquire_lingvaj_respondoj.py`

```python
#!/usr/bin/env python3
"""
Download Lingvaj Respondoj by L.L. Zamenhof.

Source: Internet Archive
License: Public Domain (Zamenhof died 1917)
"""

import requests
from pathlib import Path
import logging

ARCHIVE_URL = "https://archive.org/download/kpbc.umk.pl.Gromadzenie_POPC_024_48_247196/kpbc.umk.pl.Gromadzenie_POPC_024_48_247196.pdf"
OUTPUT_DIR = Path("data/raw/eo/lingvaj_respondoj")

def download_lingvaj_respondoj():
    """Download Lingvaj Respondoj PDF from Internet Archive."""
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    
    pdf_path = OUTPUT_DIR / "lingvaj_respondoj.pdf"
    
    if pdf_path.exists():
        logger.info(f"Already downloaded: {pdf_path}")
        return pdf_path
    
    logger.info(f"Downloading from {ARCHIVE_URL}...")
    response = requests.get(ARCHIVE_URL, stream=True)
    response.raise_for_status()
    
    with open(pdf_path, 'wb') as f:
        for chunk in response.iter_content(chunk_size=8192):
            f.write(chunk)
    
    logger.info(f"Downloaded: {pdf_path}")
    return pdf_path

def create_attribution():
    """Create README with attribution."""
    readme = OUTPUT_DIR / "README.md"
    readme.write_text("""# Lingvaj Respondoj

**Linguistic Answers by L.L. Zamenhof**

## Source Information

- **Author**: L.L. Zamenhof
- **Period**: 1889-1912
- **Editor**: Gaston Waringhien (collected edition)
- **Source URL**: https://archive.org/details/kpbc.umk.pl.Gromadzenie_POPC_024_48_247196
- **License**: Public Domain (author died 1917)

## Description

Collection of Zamenhof's authoritative answers to grammar and usage questions,
originally published in La Esperantisto, La Revuo, and La Oficiala Gazeto.

## Attribution

"Lingvaj Respondoj by L.L. Zamenhof, Public Domain"

## Usage in Klareco

- Tier 0 gold standard data (highest quality)
- Parser validation test cases
- Authoritative grammar guidance
- Disambiguation ground truth
""")
```

### Step 2: Create Cleaning Script

**File**: `scripts/clean_lingvaj_respondoj.py`

```python
#!/usr/bin/env python3
"""Extract text from Lingvaj Respondoj PDF."""

import pypdf
from pathlib import Path

def clean_lingvaj_respondoj(pdf_path: Path, output_path: Path):
    """
    Extract text from PDF, remove headers/footers.
    
    Challenges:
    - OCR quality (old PDF scan)
    - Special characters (ĉ, ĝ, ĥ, ĵ, ŝ, ŭ)
    - Headers/footers to remove
    """
    
    with open(pdf_path, 'rb') as f:
        reader = pypdf.PdfReader(f)
        
        text_parts = []
        for page in reader.pages:
            text = page.extract_text()
            # Clean headers/footers
            # Handle special characters
            text_parts.append(clean_page(text))
    
    output_path.write_text('\n\n'.join(text_parts), encoding='utf-8')
```

**Dependencies**: `pypdf` or alternative PDF extractor

### Step 3: Create Extraction Script

**File**: `scripts/extract_lingvaj_respondoj.py`

```python
#!/usr/bin/env python3
"""Extract sentences with metadata from Lingvaj Respondoj."""

def extract_lingvaj_respondoj_sentences(cleaned_text: Path, output_jsonl: Path):
    """
    Extract Q&A pairs with metadata.
    
    Structure:
    - Question about grammar
    - Zamenhof's answer/explanation
    - Example sentences
    
    Metadata:
    - Question topic
    - Date (if available)
    - Grammar rule referenced
    """
    
    with open(output_jsonl, 'w', encoding='utf-8') as out:
        for qa_pair in parse_qa_structure(cleaned_text):
            entry = {
                'text': qa_pair['answer_text'],
                'source': 'lingvaj_respondoj',
                'tier': 0,
                'weight': 10.0,
                'metadata': {
                    'question': qa_pair['question'],
                    'topic': qa_pair['topic'],
                    'date': qa_pair.get('date'),
                    'source_name': 'Lingvaj Respondoj by L.L. Zamenhof'
                }
            }
            out.write(json.dumps(entry, ensure_ascii=False) + '\n')
```

### Step 4: Shell Wrappers

```bash
#!/bin/bash
# scripts/acquire_lingvaj_respondoj.sh

./scripts/acquire_lingvaj_respondoj.py
./scripts/clean_lingvaj_respondoj.py
./scripts/extract_lingvaj_respondoj.py
```

### Step 5: Integration

Update `scripts/build_unified_corpus.py`:
```python
sources = [
    # ... existing sources
    ('lingvaj_respondoj', Path('data/extracted/lingvaj_respondoj_sentences.jsonl'), 10.0),
]
```

## Expected Output

```
data/raw/eo/lingvaj_respondoj/
├── lingvaj_respondoj.pdf (downloaded)
├── README.md (attribution)
└── LICENSE.txt (public domain notice)

data/cleaned/eo/lingvaj_respondoj/
└── lingvaj_respondoj_cleaned.txt

data/extracted/
└── lingvaj_respondoj_sentences.jsonl
    {"text": "...", "source": "lingvaj_respondoj", "tier": 0, "metadata": {...}}
```

## Success Criteria

- [ ] PDF downloaded from Internet Archive
- [ ] Attribution README created
- [ ] Text extracted from PDF (handle OCR issues)
- [ ] Q&A structure parsed
- [ ] Sentences extracted with metadata
- [ ] Integrated into unified corpus
- [ ] Parser test suite created using examples
- [ ] Models retrained with Lingvaj Respondoj data

## Dependencies

Add to `requirements.txt`:
```
pypdf>=3.0.0  # For PDF text extraction
```

## Expected Impact

- **Parser validation**: Ground truth for edge cases
- **Corpus**: +1,000-1,500 Tier 0 sentences
- **Disambiguation**: Authoritative guidance
- **Test suite**: Definitive test cases

## Estimated Effort

3-4 hours

## References

- [Internet Archive - Lingvaj Respondoj](https://archive.org/details/kpbc.umk.pl.Gromadzenie_POPC_024_48_247196)
- [Tekstaro.com version](https://tekstaro.com/t?nomo=lingvaj-respondoj)

## Labels

`enhancement`, `corpus`, `data`, `priority:P1`, `tier-0`

**Comment by marctjones** [comment]
## ✅ TEXT QUALITY VERIFIED - Source Change Recommended

### Finding: Better Source Available

**Current plan**: Download from Archive.org (OCR version)

**RECOMMENDED**: Download from **Tekstaro.com** instead

### Quality Comparison

| Source | Format | Quality | Issues |
|--------|--------|---------|--------|
| **Tekstaro.com** ✅ | Born-digital | Excellent | Zero OCR errors |
| Archive.org ⚠️ | Scanned + OCR | Variable | Potential OCR errors |

### Tekstaro.com Version Details

**URL**: https://tekstaro.com/t?nomo=lingvaj-respondoj

**Edition**: 7th edition (1990), edited by Gaston Waringhien

**Quality Features**:
- ✅ **Born-digital text** (NOT OCR)
- ✅ Clean Unicode Esperanto diacritics (ĉ, ĝ, ĵ, ŝ, ŭ)
- ✅ Editorial footnotes documenting textual variants
- ✅ Multiple correction rounds by Simon Davies
- ✅ **Most authoritative version** per Tekstaro

**Sample Quality**:
```
La vortan 'kaj' vi estas tute prava, kaj mi kore vin dankas.
```
- Perfect formatting, proper punctuation, zero character corruption

### Archive.org Version Issues

**Format**: Scanned book processed with Tesseract 5.3.0 OCR

**Potential Problems**:
- Esperanto diacritics susceptible to OCR errors (ĉ→c, ŝ→s)
- 1927 edition scan (historical typography)
- Even high confidence OCR can miss diacritical marks

### Recommendation

**Update acquisition script** to use Tekstaro.com as primary source:

```python
# scripts/acquire_lingvaj_respondoj.py

TEKSTARO_URL = "https://tekstaro.com/t?nomo=lingvaj-respondoj"

def download_from_tekstaro():
    """
    Download from Tekstaro.com (clean digital version).
    
    Options:
    1. Scrape HTML (if no download option)
    2. Check for API/export function
    3. Contact Tekstaro for bulk export
    """
```

### Benefits

- ✅ **Zero OCR errors** (Tier 0 quality preserved)
- ✅ **Most authoritative edition** (7th, corrected)
- ✅ **Parser validation ready** (perfect for #513-#516 testing)

**Full quality assessment**: /tmp/text_quality_assessment.md
